File: android/app/src/main/java/com/airbridge/app/transport/interfaces/IMessageChannel.kt
package com.airbridge.app.transport.interfaces

import com.airbridge.app.transport.protocol.ProtocolMessage
import kotlinx.coroutines.flow.Flow

/**
 * Framed message channel over a TLS TCP connection.
 * Send is safe to call from multiple coroutines concurrently.
 */
interface IMessageChannel {
    val remoteDeviceId: String
    val isConnected: Boolean

    /** Incoming message stream. Completes when the channel closes cleanly. */
    val incomingMessages: Flow<ProtocolMessage>

    suspend fun send(message: ProtocolMessage)
    suspend fun close()
}

File: windows/AirBridge.Transport/Interfaces/IMessageChannel.cs
using AirBridge.Transport.Protocol;

namespace AirBridge.Transport.Interfaces;

/// <summary>
/// Represents a framed message channel over a TLS TCP connection.
/// Handles length-prefix framing; callers work with typed <see cref="ProtocolMessage"/> objects.
/// Implementations must be safe for concurrent sends from multiple threads.
/// </summary>
public interface IMessageChannel : IAsyncDisposable
{
    string RemoteDeviceId { get; }
    bool IsConnected { get; }

    /// <summary>Sends a protocol message. Thread-safe.</summary>
    Task SendAsync(ProtocolMessage message, CancellationToken cancellationToken = default);

    /// <summary>
    /// Reads the next incoming message. Returns null if the channel is closed cleanly.
    /// Throws on transport error.
    /// </summary>
    Task<ProtocolMessage?> ReceiveAsync(CancellationToken cancellationToken = default);

    /// <summary>Raised when the connection drops unexpectedly.</summary>
    event EventHandler Disconnected;
}
